### PR TITLE
Arrangement_on_surface_2: change intersection return types (from legacy CGAL::Object to variants)

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

Change the type of the result of intersections in the entire package. More precisely, the old type was an output iterator of with value type CGAL::Object(). The new type is an output-iterator with value type variant of the possible types of results.

## Release Management

* Affected package(s): Arrangement_on_surface_2
* Issue(s) solved (if any): I don't think there was a particular issue against it
* Feature/Small Feature (if any): no
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership: TAU

I need to update the documentation as well, but the change was rather large, so I would like to run the test suite to detect possible small issues with compilers others than mine.

### CODE REVIEW

**Comment by lrineau** [comment]
Travis errors:
  - [`examples/Boolean_set_operations_2/set_union.cpp` does not compile](https://travis-ci.com/github/CGAL/cgal/jobs/311211905#L2878),
  - [`examples/Minkowski_sum_2/sum_by_decomposition.cpp` does not compile](https://travis-ci.com/github/CGAL/cgal/jobs/311211922#L2874).

**Comment by efifogel** [comment]
I've fixed these issues.
Please try again.
Thanks, Efi
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

File: Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Conic_x_monotone_arc_2.h [hunk 1]
```diff
@@ -922,22 +838,22 @@ class _Conic_x_monotone_arc_2 : private Conic_arc_
...
-                            OutputIterator oi) const
+  template <typename OutputIterator>
+  OutputIterator intersect(const Self& arc,
+                           Intersection_map& inter_map,
+                           OutputIterator oi) const
   {
-    if (_has_same_supporting_conic (arc))
-    {
+    typedef unsigned int                              Multiplicity;
+    typedef boost::variant<Intersection_point, Self>  Intersection_result;
```
Comment (lrineau) [inline Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Conic_x_monotone_arc_2.h:847]: There is a missing `#include <boost/variant.hpp>` in this file `Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Conic_x_monotone_arc_2.h`. It was detected by [our automatic tests on Travis CI](https://travis-ci.com/github/CGAL/cgal/jobs/314578509#L7177).

File: Installation/CHANGES.md [hunk 2]
```diff
@@ -6,19 +6,34 @@ Release History
...
+     new code is backward compatible. However, it is recomnded that all calls
+     to the intersection functions are fixed to use the new return type.
+
+### 2D Regularized Boolean Operations
+ -   Changed intersection return type from legacy CGAL::Object to modern
+     boost::variant.
+
+### 2D Minkowski sums
+ -   Changed intersection return type from legacy CGAL::Object to modern
+     boost::variant.
```
Comment (lrineau) [inline Installation/CHANGES.md:22]:
```suggestion
### 2D Arrangement on Surface
 -   Changed intersection return type from legacy `CGAL::Object` to modern
     `boost::variant`.
     As there is an implicit conversion from `boost::variant` to `CGAL::Object`, the
     new code is backward compatible. However, it is recommended that all calls
     to the intersection functions are fixed to use the new return type.

### 2D Regularized Boolean Operations
 -   Changed intersection return type from legacy `CGAL::Object` to modern
     `boost::variant`.

### 2D Minkowski sums
 -   Changed intersection return type from legacy `CGAL::Object` to modern
     `boost::variant`.
```

**Comment by sloriot** [comment]
```
Arrangement_on_surface_2/include/CGAL/Arr_circular_line_arc_traits_2.h:304:57: warning: typedef ‘CK’ locally defined but not used [-Wunused-local-typedefs]
Arrangement_on_surface_2/include/CGAL/Arr_circular_line_arc_traits_2.h:307:57: warning: typedef ‘Intersection_point’ locally defined but not used [-Wunused-local-typedefs]
Arrangement_on_surface_2/include/CGAL/Arr_circular_line_arc_traits_2.h:308:57: warning: typedef ‘X_monotone_curve_2’ locally defined but not used [-Wunused-local-typedefs]
Arrangement_on_surface_2/include/CGAL/Arr_geometry_traits/Conic_x_monotone_arc_2.h:847:55: warning: typedef ‘Multiplicity’ locally defined but not used [-Wunused-local-typedefs]
Boolean_set_operations_2/include/CGAL/Boolean_set_operations_2/Gps_simplifier_traits.h:185:40: warning: statement has no effect [-Wunused-value]
```

**Comment by maxGimeno** [comment]
Also `Arr_linear_traits_2.h:76:9: warning: ‘*((void*)(&<anonymous>)+4).CGAL::Arr_linear_traits_2<Kernel>::_Linear_object_cached_2::has_source’ may be used uninitialized in this function [-Wmaybe-uninitialized]`

**Comment by efifogel** [comment]
Fixed all; please re-tests...
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
warning : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-132/Arrangement_on_surface_2_Demo/TestReport_Friedrich_Ubuntu-gcc-7.gz

File: Arrangement_on_surface_2/demo/Arrangement_on_surface_2/VerticalRayShootCallback.h [hunk 3]
```diff
@@ -334,9 +346,48 @@ VerticalRayShootCallback< Arr_ >::rayShootUp( const Kernel_point_2& pt )
   return pointLocationResult;
 }
 
+template < typename Arr_ >
+CGAL::Object VerticalRayShootCallback< Arr_ >::rayShootUp( const Kernel_point_2& pt,
```
Comment (lrineau) [inline Arrangement_on_surface_2/demo/Arrangement_on_surface_2/VerticalRayShootCallback.h:350]: Was not the goal of this PR to remove usage of `CGAL::Object`?!

**Comment by efifogel** [comment]
Yes, but I also wanted to remove all warnings, and it seems that one
warning is generated while compiling the demo, so I had to fix something in
the demo in order to try and reproduce.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
I cannot reproduce the warning and it is irrelevant.
As far as I can tell the code can be integrated into master.

PS, If you really insist, we can revert the last commit;
however, since is really necessary, we will have to do it in a new
branch/pull-request right away.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by maxGimeno** [comment]
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-137.shtml

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-137.shtml

**Pull request merged**